File: week6/kotlin/factory_method_pattern/Animal.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak(): String = "Woof!"
}

class Cat : Animal() {
    override fun speak(): String = "Meow!"
}

class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

fun main() {
    val factory = AnimalFactory()

    val dog = factory.createAnimal("dog")
    val cat = factory.createAnimal("cat")

    println(dog?.speak())
    println(cat?.speak())
}

File: week6/csharp/factory_method_pattern/Animal.cs
using System;

abstract class Animal {
    public abstract string Speak();
}

class Dog : Animal {
    public override string Speak() => "Woof!";
}

class Cat : Animal {
    public override string Speak() => "Meow!";
}

class AnimalFactory {
    public Animal CreateAnimal(string animalType) {
        if (animalType == "dog") return new Dog();
        if (animalType == "cat") return new Cat();
        return null;
    }
}

class Program {
    static void Main() {
        AnimalFactory factory = new AnimalFactory();

        Animal dog = factory.CreateAnimal("dog");
        Animal cat = factory.CreateAnimal("cat");

        if (dog != null) Console.WriteLine(dog.Speak());
        if (cat != null) Console.WriteLine(cat.Speak());
    }
}
